### ISSUE TITLE
On using `xallocx` to reduce the size of an allocation in place

### ISSUE THREAD

**Issue by gnzlbg** [jemalloc/jemalloc#1222]
The `xallocx` function resizes an allocation in place. The documentation seems to be focused with "augmenting" the size of an allocation. To me, it is extremely unclear if the function can be used to reduce the size of an allocation in place at all or not. 

I'd expect that, when reducing the size of an allocation, `size` would mean "the minimum size of the allocation" and `extra` would denote the range in which the resizing succeeds or fails. 

For example, if I want to reduce a 1Gb allocation in place to 100Mb, but can live with 128Mb if the resizing happens in place, that I call `xallocx(ptr, 100Mb, 128Mb)` and that it will fail if it can only reduce the allocation from 1Gb to 800Mb in place doing as little work as possible, because if that's the case I am going to call `rallocx` afterwards anyways. 

But this is not what the docs say about `extra`, since they are focused only on growing an allocation using `xallocx`. From the docs, if `xallocx` does nothing when shrinking an allocation that's a "success" because the returned size (and size + extra) are larger than the requested one. Even if it returns a size smaller than the allocated one, the size might be larger than the one that I can accept.

**Comment by interwq** [comment]
The `xallocx` API is allowed to return size greater than `size`+`extra` -- the return value is the actual usable size.  For shrinking large allocations, the current implementation will always succeed in place and shrink to `size + extra` (unless user defined custom extent hooks disallows split). Failing a shrink request for `xallocx(ptr_to_1G, 100Mb, 128Mb)` does not sound the right thing to me.

Does that answer your question? The `extra` argument is a hint instead of a limit here.

**Comment by gnzlbg** [comment]
> Does that answer your question?

Not really. My question is, can shrinking an allocation with `xallocx` fail, and if so, how is the error reported, and if not, what is it returned?

**Comment by interwq** [comment]
Shrinking large allocations will never fail (since we simply split away the extra pages). For small allocations, since they are slab managed, even shrinking will fail unless the new size falls into the same size class. When not resized, `xallocx` returns the old size.

**Comment by gnzlbg** [comment]
So how is the error reported for small allocations?

**Comment by interwq** [comment]
Same as for growing: `The xallocx() function returns the real size of the resulting resized allocation pointed to by ptr, which is a value less than size if the allocation could not be adequately grown in place.` so you can check the returned size.

**Comment by gnzlbg** [comment]
So if the returned size is the same, the reallocation failed then.

**Comment by interwq** [comment]
That's correct.

**Comment by gnzlbg** [comment]
So say I have an allocation of size `Init` at `ptr`, and I want it to reduce it in place to a target size `TargetLow` , but can live with a size of `TargetHigh` where `TargetLow < TargetHigh < Init`. 

Let `extra = TargetHigh - TargetLow`. What's the correct way of appropriately shrinking my allocation in place?

If I perform a: `new_size = xallocx(ptr, TargetLow, extra, flags)`, IIUC, if shrinking in place to `TargetHigh` can succeed my allocation will be always shrunk to `TargetHigh` because `TargetHigh > TargetLow`. Is that correct?

If I perform a `new_size = xallocx(ptr, TargetLow, 0, flags)`, my allocation will be appropriately shrinked to `TargetLow` if that is possible, but otherwise shrinking in place will fail and `new_size == Init`. Is that correct?

If so, do I need to step from `TargetLow` up to `TargetHigh` checking whether shrinking in place succeeded? E.g. like this:

```c++
for (auto target = ranges::view::linear_distribute(TargetLow, TargetHigh, 100)) {
    size_t new_size = xallocx(ptr, target, 0, flags);
    if (new_size < Init) { break; }
}
```

**Comment by interwq** [comment]
> If I perform a: new_size = xallocx(ptr, TargetLow, extra, flags), IIUC, if shrinking in place to TargetHigh can succeed my allocation will be always shrunk to TargetHigh because TargetHigh > TargetLow. Is that correct?

> If I perform a new_size = xallocx(ptr, TargetLow, 0, flags), my allocation will be appropriately shrinked to TargetLow if that is possible, but otherwise shrinking in place will fail and new_size == Init. Is that correct? That is, the xallocx call can only return two values in this case, TargetLow or Init. Is that correct?

Both are correct. IIUC, you want it shrink to [TargetLow, TargetHigh], and as small as possible? If that's the case, try TargetLow directly for large allocations since it should succeed; and for small allocations, simply lookup the actual usable size and decide if that will work for you (since we won't be able to really resize them anyway). This may not sound ideal (and sorry to be implementation dependent) -- I agree the current API does not cover this use case.

**Comment by gnzlbg** [comment]
I think it would be good to cover this in the docs. The current API makes a lot of sense to me for growing an allocation, since one typically has a lower bound on the size, but would prefer to grow up to an upper bound if possible.

For shrinking an allocation, `size + extra` should be the largest value without failing, and if possible, `xallocx` should further shrink the allocation down to just `size`.
